### PR TITLE
WIP: Initialize spatial colors

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
see #2533

This PR will *only* address spatial colors, not topomaps.

### CODE REVIEW

**Comment by jona-sassenhagen** [comment]
@kingjr @Eric89GXL how about actually setting the light/dark red/blue gradients with a different color map?

In principle I don't mind a 2D transform (although I'm not sure it's necessary), but I'd probably need help with that.

**Comment by Eric89GXL** [comment]
@jona-sassenhagen I'm not sure how you'd do it with a colormap, seems like it would be easier to do just by setting the colors of each dot manually

**Comment by jona-sassenhagen** [comment]
Is there something specific I should know about grad pairs wrt. this? Should they be merged first or something?

**Comment by kingjr** [comment]
I would plot the same color for grad pair yes

**Comment by jona-sassenhagen** [comment]
I'm honestly not that happy with the red to blue/bright to dark map. It's actually less informative IMO because the position color codes are harder to tell apart.

This is with 2d projection, and x coordinates normalized to the .6-1 range of the Hue channel and y to the Luminosity channel.

unknown

I'd argue the original example looks a lot better.

**Comment by kingjr** [comment]
I agree, this is just all purple. 

A couple of things:
1. Do you normalize the color by axes (the sensors are not on a circle, rather on an ellipse)?
2. You may want to use a transitional color: red - green - blue instead of red/blue
3. We probably need to test the best color on a variety of contrast: left vs right hand response will be clearer than a visual response on some color mapping and vice vers.

Here are a couple of quick examples using the following code:
```

pos = find_layout(evoked.info).pos[:, :2]
pos[:, 0] = (pos[:, 0] - np.min(pos[:, 0])) / np.ptp(pos[:, 0])
pos[:, 1] = (pos[:, 1] - np.min(pos[:, 1])) / np.ptp(pos[:, 1])
cmap = plt.get_cmap('rainbow')
# red green blue: x
colors = [np.array(cmap(x)[:3]) for (x, y) in pos]
# white black: y
colors = [hls_to_rgb(*np.hstack((rgb_to_hls(*rgb)[0], y, rgb_to_hls(*rgb)[2])))
          for (rgb, y) in zip(colors, pos[:, 1])]
```

A classic mismatch field, magnetometers only:
example_mmn_topo
example_mmn

The evoked potential of a weak visual stimulus:
example_visual
example_visual_time

**Comment by agramfort** [comment]
Pretty cool !

**Comment by kingjr** [comment]
and if you wanna go fancy fancy, you could use `fill_betweenx`. I personnally use this kind of representation for significant clusters only, so that could be a different PR to twick a `mask` param
mismatch field
example_fill_mmn
weak visual response
example_fill_visual

**Comment by agramfort** [comment]
Fancy ;) yeah later.

**Comment by Eric89GXL** [comment]
-1 on red green, that is the most common color blindness

**Comment by jona-sassenhagen** [comment]
@kingjr 
>1. Do you normalize the color by axes (the sensors are not on a circle, rather on an ellipse)?

Originally, I'd just get the xyz values and directly normalize these to 0-1.
Now I do _find_topomap_coords and normalize x values to .7 to 1 (for HSV hue).

>2. You may want to use a transitional color: red - green - blue instead of red/blue
As Eric says, if we wanted to use green, we might as well stay with xyz -> RGB.

If we were to use a black, or at least very dark, background, we could use red-white-blue, as everywhere else. ColorBrewer also has a red-yellow-blue map, but I don't think it works on white either.

**Comment by jona-sassenhagen** [comment]
Here is my HSV (red-blue) code:

```
                    from matplotlib.colors import hsv_to_rgb

                    def _normalize(x):
                        x -= x.min()
                        x /= x.max()
                        return x

                    def _normalize_to_range(v, min):
                        v = _normalize(v)
                        v *= (1 - min)
                        v += min
                        return v

                    h = _normalize_to_range(xy[:, 0], .6)
                    s = [1 for _ in h]
                    v = _normalize_to_range(xy[:, 1], .3)
                    print(t, h[:3], s[:3], v[:3])

                    chspace = np.asarray([hsv_to_rgb(rgb)
                                          for rgb in zip(h, s, v)])
```

I also have a strange bug where grads are not displayed.

**Comment by jona-sassenhagen** [comment]
Argument against red-blue: topomaps are also red-blue.

**Comment by agramfort** [comment]
How does the red blue look like ?

**Comment by jona-sassenhagen** [comment]
unknown-8

Like this, right now. Not so pretty I think, and arguably less informative.

Maybe you guys have a better idea of how to do this, but I just don't think it works that well.

**Comment by kingjr** [comment]
I tend to agree with @jona-sassenhagen, red purple blue is a not sufficiently widespread color spectrum. I definitely hear the colorblind issue though; I'm just not sure whether we can have both a rich color coding and a colorblind compatible plot.
example_fill_visual
example_fill_visual

**Comment by agramfort** [comment]
I agree too. The dynamic range of the colors is limited

**Comment by kingjr** [comment]
I vote the rainbow colormap. It s a midpoint between colorblindness and
tetrachromia :]

https://en.m.wikipedia.org/wiki/Concetta_Antico

**Comment by jona-sassenhagen** [comment]
Arguably, your blue/red, light/dark thing looks a lot better than mine @kingjr .

Can we have multiple color schemes? It could just be one optional kwarg for evoked.plot(), `spatial_colors: bool | 'jet' | 'RdBu' | 'rainbow'` ...

**Comment by kingjr** [comment]
We could have a param, but because we re going for a 2d color scheme we
would need to fix the fronto posterior / black white convention. Not sure
what to think about this

**Comment by agramfort** [comment]
you need to rebase.

+1 for rainbow

when plotting with spatial_color=True it would be nice to know to where each color corresponds to.

is it possible to show this in the same figure?

**Comment by jona-sassenhagen** [comment]
Yes, like this: unknown

**Comment by agramfort** [comment]
that would be nice to put such a topo automatically in the corner of the
plot.

**Comment by jona-sassenhagen** [comment]
Yes, but I'd like to first only implement line colors so the PR doesn't grow over my head, and then make 2 more PRs for the legend and for the topoplots at the top, okay?

File: mne/viz/tests/test_evoked.py [hunk 1]
```diff
@@ -92,6 +92,8 @@ def test_plot_evoked():
         # plot with bad channels excluded
         evoked.plot(exclude='bads')
         evoked.plot(exclude=evoked.info['bads'])  # does the same thing
+        # spatial colors test
+        evoked.plot(spatial_colors=True)
```
Comment (agramfort) [inline mne/viz/tests/test_evoked.py]: maybe just put spatial_colors=True in one of the 2 plots above. it avoids test time.

**Comment by agramfort** [comment]
you need to rebase.

**Comment by agramfort** [comment]
fine

**Comment by jona-sassenhagen** [comment]
I have found something similar/prior art [here](http://jov.arvojournals.org/article.aspx?articleid=2213210):

>Averaged waveforms across all channels in the Periodic and Nonperiodic conditions (each line represents one channel and is colored in a red-to-blue gradient according to the alphabetical order of the channel name)

**Comment by agramfort** [comment]
I predict a lot of such plots in HBM in 2016 :)

**Comment by jona-sassenhagen** [comment]
@kingjr can you put up your current code as a gist?

**Comment by kingjr** [comment]
Sure

```
from colorsys import hls_to_rgb, rgb_to_hls

# normalize position to maximize color spectrum
pos = find_layout(evoked.info).pos[:, :2]
pos[:, 0] = (pos[:, 0] - np.min(pos[:, 0])) / np.ptp(pos[:, 0])
pos[:, 1] = (pos[:, 1] - np.min(pos[:, 1])) / np.ptp(pos[:, 1])
# cmap
cmap = plt.get_cmap('rainbow')
# red green blue: x-axis
colors = [np.array(cmap(x)[:3]) for (x, y) in pos]
# white black: y-axis
colors = [hls_to_rgb(*np.hstack((rgb_to_hls(*rgb)[0], y, rgb_to_hls(*rgb)[2])))
          for (rgb, y) in zip(colors, pos[:, 1])]
fig, axes = plt.subplots(1)
for chan, color in zip(evoked.data, colors):
    axes.plot(evoked.times, chan, color=color)
```

**Comment by agramfort** [comment]
you'll need to rebase

**Comment by agramfort** [comment]
here is what `evoked.plot(spatial_colors=True)` gives.

figure_1

looks nice but I don't understand why the EEG colors are less contrasted.

also we need to show the topo with colored channels as you did in lower corner.

and test on CTF data too

and demo this in an example

**Comment by jona-sassenhagen** [comment]
>looks nice but I don't understand why the EEG colors are less contrasted.

I think because here, I don't normalize channel types independently, so as the EEG channels are all closer together, there's less room for colors.

**Comment by jaeilepp** [comment]
> I think because here, I don't normalize channel types independently, so as the EEG channels are all closer together, there's less room for colors.

I did a little bit of debugging on this and the reason for the blueness was coordinate z. It was mapping xyz straight to rgb and I noticed that the z value was very high for all the EEG channels. I'm not sure about the coordinate system, but I suspect it means that most of the electrodes are above the nasion.

**Comment by jona-sassenhagen** [comment]
But shouldn't this effect disappear if all 3 dimensions are normalized to the 0, 1 range, independently for channels?
At least if I plot EEG alone, I never get this "blue tint".

**Comment by jaeilepp** [comment]
In https://github.com/mne-tools/mne-python/pull/2638 the coordinates are mapped to 2d and x and y are mapped to red and blue.

**Comment by jaeilepp** [comment]
Yeah that's true. I guess there must be just a couple of electrodes below the nasion level which will mess up the scaling.

**Comment by jona-sassenhagen** [comment]
Close for #2638

**Pull request closed without merging**